addons/crm/models/crm_lead.py

```diff
@@ -1466,20 +1479,60 @@ def _cron_update_automated_probabilities(self):
             leads_to_update = self.env['crm.lead'].search(pending_lead_domain)
-            lead_probabilities = leads_to_update._pls_get_naive_bayes_probabilities(batch_mode=True)
+            nbr_leads_to_update = len(leads_to_update)
+
+            # Compute by batch to avoid memory error
+            batch_step = 500000
+            lead_probabilities = {}
+            for i in range(0, nbr_leads_to_update, batch_step):
+                leads_to_update_part = leads_to_update[i:i+batch_step]
+                lead_probabilities.update(leads_to_update_part._pls_get_naive_bayes_probabilities(batch_mode=True))
+            _logger.info("Predictive Lead Scoring : New automated probabilities computed")
+
+            # Group by new probability to reduce server roundtrips when executing the update
+            lead_ids_by_probabilities = defaultdict(list)
+            for lead_id, proba in sorted(lead_probabilities.items()):
+                lead_ids_by_probabilities[proba].append(lead_id)
 
-            # Update in execute batch to avoid server roundtrips + page_size to 10000 to avoid memory errors
             # Update both probability and automated_probability if they were equal, else, update only automated_probability
             sql = """UPDATE crm_lead
                         SET automated_probability = %s,
-                            probability = CASE WHEN (ROUND(probability::numeric, 2) = ROUND(automated_probability::numeric, 2) or probability is null)
+                            probability = CASE WHEN (probability = automated_probability or probability is null)
                                                THEN (%s)
                                                ELSE (probability)
                                                END
-                        WHERE id = %s"""
-            batch_params = [(lead_probabilities[lead.id], lead_probabilities[lead.id], lead.id) for lead in leads_to_update if lead.id in lead_probabilities]
-            extras.execute_batch(self._cr, sql, batch_params, page_size=10000)
-            _logger.info("Predictive Lead Scoring : all automated probability updated (count: %d)" % (len(leads_to_update)))
+                        WHERE id in %s"""
+
+            # Update by a maximum number of leads at the same time, one batch by transaction :
+            # - avoid memory errors
+            # - avoid blocking the table for too long with a too big transaction
+            batch_step = 5000
+            treated_leads = 0
+            nbr_of_transactions = 0
+            cron_update_lead_start_date = datetime.now()
+            auto_commit = not getattr(threading.currentThread(), 'testing', False)
+            for proba, lead_ids in lead_ids_by_probabilities.items():
+                for idx_from in range(0, len(lead_ids), batch_step):
+                    lead_ids_to_exec = lead_ids[idx_from:idx_from + batch_step]
+                    nbr_of_transactions += 1
```

Comment:
Is it usefull ? Doing len(lead_ids) / batch_step is the same, if only for reporting.

Reply:
not really the same, because if a probability has more related elements than the batch_step, it can make multiple transactions
example : `step = 5, probability_leads= {23: [1,2,3,4,5,6,7,8,9,10,11]}`
--> 3 transaction for 1 proba. (in this case this works).
but if : 
```
probability_leads= {
    23: [1],
    24: [2],
    25: [3],    
    26: [4],
    27: [5],
    28: [6],
    29: [7],
}
```
we get 7 transactions, instead of 3.
and combination of both is quite complex. :D